OpenAPITools/openapi-generator
# [csharp] use full qualified name to fix cases where generated code doesn't build

This fixes #16123

The issue is that when a spec defines an entity with a name like `TimeZone`, the generated code does not build because it yields an ambiguous reference to either our entity or `System.TimeZone`.
That commit fixes it by adding the namespace in front of the entity name.
However, it adds the suffix only if need, ie: if we consider that the http call will return an `Object` or a `string`, then adding our package namespace would break the generated code. That commit hence ensures it does not do that.


To test this patch I:

* created that spec file

```
openapi: 3.0.3
info:
 title: test
 version: v0.1
 description: test
paths:
 /api/path1: # This path makes sure we fix our issue, because it would yield C# code that does not build without this commit
  get:
   responses:
    '200':
      description: ok
      content:
       application/json:
        schema:
         $ref: '#/components/schemas/TimeZone'
 /api/path2: # This path leads to C# code that manipulates an Object. This ensure I don't break support of that case.
  get:
   responses:
    '200':
      description: ok
      content:
       application/json:
        schema:
         $ref: '#/components/schemas/ObjectIsh'
 /api/path3: # This path leads to C# code that manipulates a string. This ensure I don't break support of that case.
  get:
   responses:
    '200':
      description: ok
      content:
       application/json:
        schema:
         $ref: '#/components/schemas/StringIsh'
components:
  schemas:
   TimeZone:
    type: object
    properties:
      someField:
       type: array
       items:
         type: string
   ObjectIsh:
    type: object
   StringIsh:
    type: string
```

I then generate C# code from this spec:

    java -jar openapi-generator-cli.jar generate --generator-name csharp --input-spec spec.yaml --output out

and then I build the generated code:

    dotnet build Org.OpenAPITools.sln

Expected behavior: the C# build is successful.
Actual behavior: without this commit the build fails. With this commit it succeeds


Ping @mandrean , @shibayan , @Blackclaws , @lucamazzanti , @muttleyxd 




### PR checklist
 
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [ ] In case you are adding a new generator, run the following additional script : 
  ```
  ./bin/utils/ensure-up-to-date
  ``` 
  Commit all changed files.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
